File: packages/backend/app/graphql/mutations/create_admin_word.rb
module Mutations
  # 公式単語帳への単語の追加（管理者専用）。教材・章どちらの公式単語帳にも追加できる。
  # 公式単語は user を持たない（Word#user_consistency）ため user は付けない。
  class CreateAdminWord < BaseAdminWordMutation
    argument :wordbook_id, ID, required: true, description: "追加先の公式単語帳（教材・章）"
    argument :question, String, required: true
    argument :answer, String, required: true

    def resolve(wordbook_id:, question:, answer:)
      return failure(forbidden_errors) unless current_admin

      wordbook = Wordbook.official.kept.find_by(id: wordbook_id)
      return failure(wordbook_not_found_errors) unless wordbook

      word = wordbook.words.build(question: question, answer: answer)

      return failure(validation_errors(word)) unless word.valid?

      word.save!

      { success: true, errors: [], word: word }
    end

    private

    def wordbook_not_found_errors
      [ { field: "wordbookId", message: "追加先の公式単語帳が見つかりません" } ]
    end
  end
end

File: packages/frontend/src/graphql/mutations/createAdminWord.ts
"use client";

import { useMutation } from "@apollo/client/react";
import { graphql } from "@/gql";
import type { CreateAdminWordMutationVariables } from "@/gql/graphql";

/** 公式単語帳への単語の追加（管理者専用）。 */
export const CreateAdminWordDocument = graphql(`
  mutation CreateAdminWord($wordbookId: ID!, $question: String!, $answer: String!) {
    createAdminWord(wordbookId: $wordbookId, question: $question, answer: $answer) {
      success
      errors {
        field
        message
      }
      word {
        id
        question
        answer
      }
    }
  }
`);

export type { CreateAdminWordMutationVariables };

export function useCreateAdminWordMutation() {
  return useMutation(CreateAdminWordDocument);
}
